source/crud/tests/README.rst

```diff
@@ -90,10 +100,155 @@ Each YAML file has the following keys:
+
+  - If the top-level ``data`` field is present in the test file, insert the
+    corresponding data into the collection under test using
+    ``globalMongoClient``.
+
+  - If the the ``failPoint`` field is present, use ``globalMongoClient`` to
+    configure the fail point on the primary server. See
+    `Server Fail Point <../../transactions/tests#server-fail-point>`_ in the
+    Transactions spec test documentation for more information.
+
+  - Create a local MongoClient (``localMongoClient``) and connect to the server.
+    This client will be used for executing the test case.
+
+      - If ``clientOptions`` is present, those options should be used to create
+        the client. Drivers MAY merge these options atop existing defaults (e.g.
+        reduced ``serverSelectionTimeoutMS`` value for faster test failures) at
+        their own discretion.
+
+  - Activate command monitoring for ``localMongoClient`` and begin capturing
+    events. Note that some events may need to be filtered out if the driver
+    uses global listeners or reports internal commands (e.g. ``isMaster``,
+    authentication).
+
+  - For each element in the ``operations`` array:
+
+    - Using ``localMongoClient``, select the appropriate ``object`` to execute
+      the operation. Default to the collection under test if this field is not
+      present.
+
+      - If ``collectionOptions`` is present, those options should be used to
+        construct the collection object.
+
+    - Given the ``name`` and ``arguments``, execute the operation on the object
+      under test. Capture the result of the operation, if any, and observe
+      whether an error occurred. If an error is encountered that includes a
+      result (e.g. BulkWriteException), extract the result object.
+
+    - If ``error`` is present and true, assert that the operation encountered an
+      error. Otherwise, assert that no error was encountered.
+
+    - if ``result`` is present, assert that it matches the operation's result.
+
+  - Deactivate command monitoring for ``localMongoClient``.
+
+  - If the ``expectations`` array is present, assert that the sequence of
+    emitted CommandStartedEvents from executing the operation(s) matches the
+    sequence of ``command_started_event`` objects in the ``expectations`` array.
+
+  - If the ``outcome`` field is present, assert the contents of the specified
+    collection using ``globalMongoClient``.
+
+Evaluating Matches
+------------------
+
+The expected values for ``result``, ``command_started_event.command``, and
+elements in ``outcome.data`` are written in
+`Extended JSON <../../extended-json.rst>`_. Drivers may adopt any of the
+following approaches to comparisons, as long as they are consistent:
+
+- Convert ``actual`` to Extended JSON and compare to ``expected``
+- Convert ``expected`` and ``actual`` to BSON, and compare them
+- Convert ``expected`` and ``actual`` to native representations, and compare
+  them
+
+Extra Fields in Actual Results
```

Comment:
Does this apply for arrays as well? For example, would:

- ``expected`` is `[1, 4, 5]`
- ``actual`` is `[1, 4, 5, 6]`

be considered a match?

Reply:
No sir. I'll clarify that this is only for extra fields in documents (e.g. `command_started_event.command`, `outcome.data[i]`, and documents within result objects).

Reply:
See https://github.com/mongodb/specifications/pull/453/commits/517d8510340fc225b4e61a790e098869aef528a8